File: Regular Expression Matching.cpp
class Solution {
public:
    bool isMatch(const char *s, const char *p) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function   
        if (p[0] == '\0') return s[0] == '\0';
        
        if (p[1] != '*') {
            return (s[0] == p[0] || p[0] == '.' && s[0] != '\0') && isMatch(s+1, p+1);
        } else {
            return (s[0] == p[0] || p[0] == '.' && s[0] != '\0') && isMatch(s+1, p) || isMatch(s, p+2);
        }
    
    }
};


class Solution {
public:
    bool isMatch(const char *s, const char *p) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function    
        if (*p == '\0') return *s == '\0';
        
        if (*(p+1) != '*') {
            return (*p == *s || (*p == '.' && *s != '\0')) && isMatch(s+1, p+1);
        }
        
        while (*p == *s || (*p == '.' && *s != '\0')) {
            if (isMatch(s, p+2)) return true;
            s++;
        }
        return isMatch(s, p+2);
    }
};

File: Regular Expression Matching.cs
public class Solution {
    public bool IsMatch(string s, string p) {
        return IsMatch(s, p, 0, 0);
    }
    
    private bool IsMatch(string s, string p, int i, int j) {
        if (j == p.Length) {
            return i == s.Length;
        }
        
        
        if (j < p.Length - 1 && p[j + 1] == '*') {
            return (i < s.Length && s[i] == p[j] || p[j] == '.' && i != s.Length) && IsMatch(s, p, i + 1, j) || IsMatch(s, p, i, j + 2);
        } else {
            return (i < s.Length && s[i] == p[j] || p[j] == '.' && i != s.Length) && IsMatch(s, p, i + 1, j + 1);
        }
    }
}
